cpp/benchmarks/ndsh/utilities.cpp

```diff
@@ -356,24 +360,32 @@ void write_to_parquet_device_buffer(std::unique_ptr<cudf::table> const& table,
 
   // Write parquet data to host buffer
   auto builder =
-    cudf::io::parquet_writer_options::builder(cudf::io::sink_info(&h_buffer), table->view());
+    cudf::io::parquet_writer_options::builder(source.make_sink_info(), table->view());
   builder.metadata(table_input_metadata);
   auto const options = builder.build();
-  cudf::io::write_parquet(options);
+  cudf::io::write_parquet(options, stream);
+  // stream.synchronize();
+}
 
-  // Copy host buffer to device buffer
-  source.d_buffer.resize(h_buffer.size(), stream);
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    source.d_buffer.data(), h_buffer.data(), h_buffer.size(), cudaMemcpyDefault, stream.value()));
+inline auto make_managed_pool()
+{
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(
+    std::make_shared<rmm::mr::managed_memory_resource>(), rmm::percent_of_free_device_memory(50));
 }
 
 void generate_parquet_data_sources(double scale_factor,
                                    std::vector<std::string> const& table_names,
-                                   std::unordered_map<std::string, parquet_device_buffer>& sources)
+                                   std::unordered_map<std::string, cuio_source_sink_pair>& sources)
 {
+  // Set the memory resource to the managed pool
+  auto old_mr = cudf::get_current_device_resource(); // fixme: already pool takes 50% of free memory
+  // TODO: release it, and restore it later?
+  auto managed_pool_mr = make_managed_pool();
+  cudf::set_current_device_resource(managed_pool_mr.get());
```

Comment:
Why not pass the new mr all the way through instead of resetting the current one? 
All the libcudf APIs should take an `mr` parameter now.

Reply:
Thanks @davidwendt. I've asked @karthikeyann to separate the MR used for data generation versus the one used for timed query runs. We need managed memory to avoid OOM in the generator, but we mostly care about async and pool for timed runs.

Reply:
@davidwendt This brings us to an old question. Right now, we can only pass `mr` for output values of a libcudf function. All intermediate allocations happen using `cudf::get_current_device_resource()`. So, if we are targeting larger than GPU memory, the intermediate allocations might run out of GPU memory if the `cudf::get_current_device_resource()` is not managed memory. Right now, libcudf functions does not have a way pass an `mr` for intermediate allocations. It's set globally using `cudf::set_current_device_resource`. That's the reason why `cudf::set_current_device_resource` is updated here.

Reply:
Ok, I thought that may be the case but wanted to make sure. It seems like we should purposely pass an `mr` for the returned objects if only for illustration purposes to highlight there are 2 mrs in play here.
Also, it may be worth a detailed comment in the code similar to what you responded here.

Reply:
I will add it to comments.
If we have a way to solve intermediate allocations, that would be great.

Reply:
Also, `old_mr` takes some memory (50% is default) already if it is a pooled memory resource. So, it will cause managed memory resource to spill more often. It's better if that also can be reclaimed until data generation is over. I am still working on this part; a shrink function in pooled memory resource would be great, but not available right now. releasing pool memory would be dangerous (all old allocations become dangling pointers). I am looking into other memory resources for this.

Reply:
You could just request these benchmarks be run using a different default pool (command-line parameter `rmm_mode`) to start with. These benchmarks are not run in CI. I'm not sure it is worth circumventing the pool since that logic would need to account for the parameter setting the default to not-pool in any case.
https://github.com/rapidsai/cudf/blob/7173b52fce25937bb69e22a083a5de4655078fa1/cpp/benchmarks/fixture/nvbench_fixture.hpp#L126
Perhaps you could even check the default `mr` somehow and if it is set to `pool` then throw an exception.

Reply:
I updated the code to create managed pool, if the existing mr is not managed or managed_pool.
It has a drawback when `pool` memory is used, which is default. and data generation may be slow. If that's not acceptable, only way we could fix it is by creating new nvbench_fixture for ndsh benchmarks alone.

@GregoryKimball can we limit rmm_mode to be managed/managed_pool only for running these benchmarks? if we can limit to managed_pool only, no fix required for mr, just use managed_pool or managed mode in cli. Alternatively, if rmm_mode could be anything, and we still want the data generator to be fastest, we should fix this PR by creating a new nvbench_fixture.

Reply:
@karthikeyann and I had an offline discussion about the NDS-H-cpp benchmarks. We agreed to collect some data on the max scale factor for the simple single-MR case and the more complex two-MR case.